packages/rocketchat-livechat/server/hooks/saveAnalyticsData.js

```diff
@@ -0,0 +1,45 @@
+RocketChat.callbacks.add('afterSaveMessage', function(message, room) {
+	// skips this callback if the message was edited
+	if (!message || message.editedAt) {
+		return message;
+	}
+
+	// check if room is livechat
+	if (room.t !== 'l') {
+		return message;
+	}
+
+	Meteor.defer(() => {
+		const now = new Date();
+		let analyticsData;
+
+		// if the message has a token, it was sent by the visitor
+		if (!message.token) {
+			const visitorLastQuery = (room.metrics && room.metrics.v) ? room.metrics.v.lq : room.ts;
+			const agentLastReply = (room.metrics && room.metrics.servedBy) ? room.metrics.servedBy.lr : room.ts;
+			const agentJoinTime = (room.servedBy && room.servedBy.ts) ? room.servedBy.ts : room.ts;
+
+			if (agentLastReply === room.ts) {		// first response
+				analyticsData = {};
+				analyticsData.firstResponseDate = now;
+				analyticsData.firstResponseTime = (now.getTime() - visitorLastQuery) / 1000;
+				analyticsData.responseTime = (now.getTime() - visitorLastQuery) / 1000;
+				analyticsData.avgResponseTime = (((room.metrics && room.metrics.response && room.metrics.response.tt) ? room.metrics.response.tt : 0) + analyticsData.responseTime) / (((room.metrics && room.metrics.response && room.metrics.response.total) ? room.metrics.response.total : 0) + 1);
+
+				analyticsData.firstReactionDate = now;
+				analyticsData.firstReactionTime = (now.getTime() - agentJoinTime) / 1000;
+				analyticsData.reactionTime = (now.getTime() - agentJoinTime) / 1000;
+			} else if (visitorLastQuery > agentLastReply) {		// response, not first
+				analyticsData = {};
+				analyticsData.responseTime = (now.getTime() - visitorLastQuery) / 1000;
+				analyticsData.avgResponseTime = (((room.metrics && room.metrics.response && room.metrics.response.tt) ? room.metrics.response.tt : 0) + analyticsData.responseTime) / (((room.metrics && room.metrics.response && room.metrics.response.total) ? room.metrics.response.total : 0) + 1);
+
+				analyticsData.reactionTime = (now.getTime() - visitorLastQuery) / 1000;
+			}	// ignore, its continuing response
+		}
+
```

Comment:
You can improve all of this code.. you just need to create variables instead of props, like this:
```
let firstReactionDate;
let firstReactionTime;
```

At the end of the process you can use the shorthand properties like below:

```
const analyticsData = {
...
firstReactionDate,
firstReactionTime,
...
}
```